lib/BackbeatConsumer.js

```diff
@@ -17,6 +18,16 @@ const { observeKafkaStats } = require('./util/probe');
 
 // controls the number of messages to process in parallel
 const CONCURRENCY_DEFAULT = 1;
+// controls the max number of messages to queue for processing.
+// messages with the same key are queued to ensure they are
+// processed in order. We should avoid having a high value for
+// this parameter as it may lead to:
+// - Memory exhaustion if the queue is not drained fast enough
+// - Higher risk of processing previously consumed messages in case
+// of a crash or rebalance, as the offsets are committed strictly in
+// order, and the queueing of messages blocks the next offsets for
+// the same partition from being commited.
+const MAX_QUEUED_DEFAULT = 1000;
```

Comment:
should this be a "total" value, or a per-queue limit (i.e. total max queued items would be `concurrency x maxQueueLength`) ?

Reply:
I think having a per-queue limit would make the code more complex for little practical benefit in prod. Having that would mean tracking the length of each of the queues, blocking when one gets full or otherwise read what we can. Can be done, but not sure it's worth it.

Reply:
On the value 1000, was it chosen arbitrary or do we have a reason not to have higher numbers? (not very high, so we don't risk re processing events in case of a crash)

Reply:
I avoided setting a high value because of the reason you stated yes, i.e with a high value there would be a higher risk of processing previous messages in case of a restart or rebalance as the queued messages will block the offset to be committed (offset commits are strictly done in order)

In practice, this value should be high enough as it's unlikely that we would be replicating/transitioning/expiring a large number of versions of the same object at a time.

Reply:
thanks!